### DISCUSSION TITLE
Profiling

### DISCUSSION THREAD

**Discussion by kellytk** [tokio-rs/axum#1322] [Q&A]
Which profiling tooling, ideally written in Rust itself, is best for use with Axum-based apps?

**Comment by davidpdrsn** [comment]
I don't know. Haven't done much profiling outside of benchmarks with rewrk (dunno if that's considered profiling)

**Comment by yanns** [comment]
Here is a list of tools I've used:
- on mac: https://github.com/cmyr/cargo-instruments
- on linux:`perf`

I've also used https://pyroscope.io/ inside a Kubernetes cluster. They have a rust sdk that you can integrate inside your application: https://grafana.com/docs/pyroscope/latest/configure-client/language-sdks/rust/

and I run benchmarks with https://github.com/flamegraph-rs/flamegraph
